util/metrics-service/src/config.rs

```diff
@@ -0,0 +1,191 @@
+
+use metrics_core::Observe;
+use metrics_exporter_http::HttpExporter;
+use metrics_exporter_log::LogExporter;
+use metrics_observer_json::JsonBuilder;
+use metrics_observer_prometheus::PrometheusBuilder;
+use metrics_observer_yaml::YamlBuilder;
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+
+use ckb_async_runtime::Handle;
+
+/* Examples:
+ * ```toml
+ * [metrics]
+ * threads = 3
+ * histogram_window = 60
+ * histogram_granularity = 1
+ * upkeep_interval = 500
+ * [metrics.exporter.prometheus]
+ * target = { type = "http", listen_address = "127.0.0.1:8100" }
+ * format = { type = "prometheus" }
+ * [metrics.exporter.log_yaml]
+ * target = { type = "log", level = "warn", interval = 600 }
+ * format = { type = "yaml" }
+ * [metrics.exporter.log_json]
+ * target = { type = "log", level = "error", interval = 900 }
+ * format = { type = "json" }
+ * ```
+ */
+#[derive(Default, Clone, Debug, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Config {
+    #[serde(default)]
+    pub threads: usize,
+
+    #[serde(default)]
+    pub histogram_window: u64, // seconds
+    #[serde(default)]
+    pub histogram_granularity: u64, // seconds
+    #[serde(default)]
+    pub upkeep_interval: u64, // milliseconds
+
+    #[serde(default)]
+    pub exporter: HashMap<String, Exporter>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Exporter {
+    pub target: Target,
+    pub format: Format,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+#[serde(tag = "type")]
+pub enum Target {
+    Log {
+        level: LogLevel,
+        interval: u64, // seconds
+    },
+    Http {
+        listen_address: String,
+    },
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+#[serde(tag = "type")]
+pub enum Format {
+    Json {
+        #[serde(default)]
+        pretty: bool,
+    },
+    Yaml,
+    Prometheus,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum LogLevel {
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+impl Config {
+    // .map_err(|err| format!("metrics exporter {}: {}", name, err))?;
+
+    pub fn check_exporter_name(name: &str) -> Result<(), String> {
+        if name.is_empty() {
+            return Err("the name of exporter shouldn't be empty".to_owned());
+        }
+        match Regex::new(r"^[0-9a-zA-Z_-]+$") {
+            Ok(re) => {
+                if !re.is_match(&name) {
+                    return Err(format!(
+                        "invaild exporter name \"{}\", only \"0-9a-zA-Z_-\" are allowed",
+                        name
+                    ));
+                }
+            }
+            Err(err) => {
+                return Err(format!("failed to check the name of exporter: {}", err));
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Exporter {
+    pub(crate) fn run<C>(self, handle: &Handle, c: C) -> Result<(), String>
```

Comment:
Just a question, why not put the function `run` into "service.rs".

Reply:
- **`Exporter` is defined in this file.** And 90% code in `run` is to parse the configuration (just as the file name `config.rs`).
- At first, I want to return a generic `MetricsExporter` from the `ConfigExporter`, then I can `async_run()` the generic `MetricsExporter` in "service.rs".
  **But**, I failed.
  Ideally, the code should be:
  ```rust
  impl ConfigExporter {
      pub(crate) fn into_metrics_exporter(self, observe) -> MetricsExporter {
          let Self { target, format } = self;
          let builder = Builder::from(format);
          let exporter = MetricsExporter::new(observe, builder, target);
          return exporter;
      }
  }

  pub fn init(config) {
      /* ... omit ... */
      for (name, exporter) in config.exporter {
          Config::check_exporter_name(&name)?;
          let metrics_exporter = exporter.into_metrics_exporter();
          handle.spawn(async {
              tokio::spawn(metrics_exporter.async_run(controller.clone()));
          });
      }
      /* ... omit ... */
  }
  ```

Reply:
As I suggest that config should be its own crate, the `run` method should be in the service crate instead in the config crate.